README.md
# prettybreaks

`prettybreaks` contains functions to create pretty breaks for numeric vectors. 

Its main function `pretty2()` is similar to `base::pretty()` but allows to use different number systems and other tweaks.

## Actual goal

The package is very simple by purpose. Its main aim is to give a good starting point for building a new package, see file "packaging.R" for the full workflow.

## Installation

``` r
library(devtools)
install_github("mayer79/prettybreaks")
```

## Teaser

``` r
library(prettybreaks)

x <- 1:100

pretty2(x)                           # c(0, 20, 40, 60, 80, 100)
pretty2(x, n = 4)                    # c(0, 50, 100, 150)
pretty2(x, base = 5)                 # c(0, 25, 50, 75, 100)
pretty2(x, p = c(10/7, 20/7, 50/7))  # c(0., 28.57143, 57.14286, ...)
```



NEWS.md
# prettybreaks 0.0.1

This is the initial release.


tests/testthat.R
library(testthat)
library(prettybreaks)

test_check("prettybreaks")


tests/testthat/test-pretty2.R
context("Unit tests for pretty2")

test_that("default arguments lead to expected output", {
  expect_equal(pretty2(0:100), c(0, 20, 40, 60, 80, 100))
})

test_that("the first argument should not be constant", {
  expect_error(pretty2(c(0, 0, 0)))
})

# Add many additional tests. For a small package, one test file is sufficient.
# For larger packages, more test scripts are recommended.


packaging.R
#===============================================================================
# Create R package
#
# This script creates a simple R package with optional vignette and unit tests.
# Much more info on https://r-pkgs.org/
#===============================================================================

# WORKFLOW FOR CREATING A NEW PACKAGE
# 1) Clone this template from https://github.com/mayer79/prettybreaks.
# 2) Rename main folder to YOUR package name, remove folder ".git" and
#    create a fresh git tracking (in RStudio or by running "git init").
# 3) If you don't need them, delete folders "tests" and/or "vignettes".
# 4) Replace word "prettybreaks" in NEWS.md, packaging.R, *.Rproj,
#    README.md, tests/, vignettes/ by your package name.
# 5) Put user visible functions of your package into folder "R" as
#    [function_name].R each. Helper functions can be added (undocumented) to
#    "R/utils.R" or in any of the existing scripts in folder "R".
# 6) Edit NEWS.md, README.md, packaging.R and the content of the optional
#    folders "tests" and "vignette" to describe your package.
#    If the package should go to CRAN, edit file "cran-comments.md" as well.
# 7) Restart RStudio and run this script. It will make a zip and tar.gz package.
#    The vignette is not part of the package but can be found in folder "doc".

# WORKFLOW TO UPDATE PACKAGE TO NEW VERSION
# 1) Update functions, documentation (NEWS.md, README.md, vignettes, tests,
#    and - if package is on CRAN - also cran-comments.md).
# 2) Increase package version in the script below and run this script.

# NOTES
# - Should something go wrong, fix the problems and rerun this script.
# - In your scripts in folder "R", never load a library or run "source".
#   The use of "packagename::function" is fine, while you can't use
#   internal functions like "packagename:::function".
# - Keep the dependency footprint small, i.e., don't use unnecessary packages.
# - If the package should go to CRAN, no warnings/notes should appear in check().
# - Git tracking is recommended, but not strictly required.

library(usethis)
library(devtools)
library(testthat)
library(knitr)
library(rmarkdown)

has_tests <- dir.exists("tests")
has_vignette <- dir.exists("vignettes")

# Clean up some files and folders. They will be generated later.
for (f in c(".Rbuildignore", "DESCRIPTION", "NAMESPACE", "LICENSE.md", "R/utils-pipe.R"))
  if (file.exists(f)) file.remove(f)
for (f in c("doc", "Meta", "man"))
  if (dir.exists(f)) unlink(f, recursive = TRUE)

# Create description file
use_description(
  fields = list(
    Package = "prettybreaks",
    Title = "Pretty Breaks",
    Version = "0.0.1",
    Date = Sys.Date(),
    Description = "Creation of pretty breaks for numeric vectors, extending the base function `pretty()` in R base.",
    `Authors@R` = "person('Michael', 'Mayer', email = 'mayermichael79@gmail.com', role = c('aut', 'cre'))",
    Depends = "R (>= 3.1.0)",
    VignetteBuilder = "knitr",  # maybe remove if there is no vignette
    LazyData = NULL,            # change to TRUE if there is data in your package
    Maintainer = "Michael Mayer <mayermichael79@gmail.com>"),
  roxygen = TRUE)
use_gpl_license(2)  # To avoid a warning. For private packages, this is a bit strange.
# use_github_links() # use this if this project is public(!) and is on github

# List required external packages (one row per package)
# use_package("stats", "Imports")  # If your code uses functions from "stats"
# use_pipe()                       # if your code uses the pipe operator %>%

# Packages used for unit tests and/or vignettes
if (has_tests) use_package("testthat", "Suggests")
if (has_vignette) {
  use_package("knitr", "Suggests")
  use_package("rmarkdown", "Suggests")
}

# List files and folders that do not belong to the package
ignore <- c(".git", ".gitignore", "packaging.R", "cran-comments.md",
            "doc", "Meta", "[.]Rproj$", ".Rproj.user")
use_build_ignore(ignore, escape = FALSE)

# Should you need to include data in your package
# use_data(...) # and document it in R/data.R

# Finish package building
document()                           # Create documentation from Roxygen tags (@param etc.)
if (has_tests) test()                # Run unit tests
if (has_vignette) build_vignettes()  # Build vignette (it will be put into "doc/")
check(vignettes = FALSE)             # Run package checks. No errors or warnings should be left.
build()                              # Create tag.gz package in parent(!) folder
build(binary = TRUE)                 # Create zip (for useres without RTools)
install()                            # Install package on your machine

# Run only if package is public(!) and should go to CRAN
if (FALSE) {
  check_win_devel()
  check_rhub()

  # Wait until above checks are passed without relevant notes/warnings
  # then submit to CRAN
  devtools::release()
}



R/pretty2.R
#' Pretty Breakpoints
#'
#' Classic algorithm to create pretty breaks from a numeric vector, see reference. In contrast to \code{base::pretty()}, this function works for different number systems and with different notion of "pretty" numbers.
# #' @importFrom stats aggregate, ave  # Import some functions from a package
# #' @import ggplot2                   # Import package if many functions needed
#' @param x Numpy array of size >= 2. Used to derive breaks.
#' @param n Approximate number of intervals between breaks. Defaults to 5.
#' @param p List of basic rounding numbers in [1, base), e.g., \code{p = 10/7} will lead to multiples of 1/7, 10/7, 100/7 etc, whatever fits best to \code{x} and \code{n}. Defaults to \code{c(1, 2, 5)}.
#' @param base Radix of the number system >= 2. Defaults to 10.
#' @param tol Numeric tolerance for close 0 breaks. Defaults to 1e-9.
#' @return A vector with breaks.
#' @export
#' @references
#' W. J. Dixon and R. A. Kronmal (1965). The choice of origin and scale for graphs. Journal of the ACM, 12(2):259-261.
#' @examples
#' x <- 1:100
#'
#' pretty2(x)                           # c(0, 20, 40, 60, 80, 100)
#' pretty2(x, n = 4)                    # c(0, 50, 100, 150)
#' pretty2(x, base = 5)                 # c(0, 25, 50, 75, 100)
#' pretty2(x, p = c(10/7, 20/7, 50/7))  # c(0., 28.57143, 57.14286, ...)
pretty2 = function(x, n = 5, p = c(1, 2, 5), base = 10, tol = 1e-9) {
  # Initialization and checks
  stopifnot(length(x) >= 2L, base >= 2, n >= 2)

  A <- min(x)
  B <- max(x)
  R <- B - A

  if (R <= 0) {
    stop("x is a constant vector!")
  }

  n <- trunc(n)

  p <- sort(p)
  p <- p[(p >= 1) & (p < base)]

  if (length(p) < 1L) {
    stop("p must contain at least one number in [1, base).")
  }

  ti <- floor(logb(R / n, base = base))
  k <- ti + (R / n / base^ti > max(p))
  i <- which(p >= R / n / base^k)[1]

  # Until scale is large enough
  b <- B - 1
  while (B > b) {
    s <- p[i] * base^k
    a <- s * floor(((B + A) / 2 - s * n / 2) / s)
    b <- a + s*n

    if (i < n) {
      i <- i + 1
    } else {
      i <- 1
      k <- k + 1
    }
  }

  # Organize output
  out <- seq(a, b, s)
  out <- out[(out >= A - s) & (out <= B + s)]
  out[abs(out) < tol] <- 0
  out
}

# Put (undocumented) helper functions here

